test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ControllerTest.cs

```diff
@@ -1797,34 +1779,41 @@ public void TempData_CanSetAndGetValues()
             Assert.Equal(input, result);
         }
 
-        private static Controller GetController(IModelBinder binder, IValueProvider provider)
+        private static Controller GetController(IModelBinder binder, IValueProvider valueProvider)
         {
             var metadataProvider = TestModelMetadataProvider.CreateDefaultProvider();
             var httpContext = new DefaultHttpContext();
-            var actionContext = new ActionContext(httpContext, new RouteData(), new ActionDescriptor());
 
             var viewData = new ViewDataDictionary(metadataProvider, new ModelStateDictionary());
             var tempData = new TempDataDictionary(Mock.Of<IHttpContextAccessor>(), Mock.Of<ITempDataProvider>());
 
-            var bindingContext = new ActionBindingContext()
+            var controllerContext = new ControllerContext()
             {
-                ModelBinder = binder,
-                ValueProvider = provider,
+                HttpContext = httpContext,
                 InputFormatters = new List<IInputFormatter>(),
-                ValidatorProvider = new DataAnnotationsModelValidatorProvider(
-                    options: null,
-                    stringLocalizerFactory: null)
+                ModelBinders = new List<IModelBinder>()
```

Comment:
The type of the array matches the type of the items inside `{ ... }` if they all match exactly. This would work in the cases you're suggesting. I usually avoid this pattern because it breaks if you add another thing that isn't typed exactly the same. 

It don't believe it tries to generate a 'bottom' type when inferring the value of `new [] { ... }`

OK
```
controller.ControllerContext.ModelBinders = new []{ (IModelBinder)foo, (IModelBinder)bar, };
```

Un-OK
```
controller.ControllerContext.ModelBinders = new []{ (MyModelBinder1)foo, (MyModelBinder2)bar, };
```